src/megaclient.cpp

```diff
@@ -13705,19 +14081,269 @@ void MegaClient::execmovetosyncdebris()
+{
+    error e = API_OK;
+
+    if (!syncConfigs || !syncConfigs->removeByTag(tag))
+    {
+        LOG_err << "Found no config for tag " << tag << " upon sync removal";
+        e = API_ENOENT;
+    }
+
+    app->sync_removed(tag);
+    return e;
+}
+
+error MegaClient::removeSyncConfigByNodeHandle(mega::handle nodeHandle)
+{
+    auto tag = 0;
+
+    if (!syncConfigs)
+    {
+        LOG_err << "no SyncConfig upon removeSyncConfigByNodeHandle";
+        return API_ENOENT;
+    }
+
+    auto config = syncConfigs?syncConfigs->getByNodeHandle(nodeHandle):nullptr;
+    if (config)
+    {
+        tag = config->getTag();
+        return removeSyncConfig(config->getTag());
+    }
+
+    LOG_err << "Found no config for handle " << nodeHandle << " upon sync removal";
+    return API_ENOENT;
+}
+
+error MegaClient::saveAndUpdateSyncConfig(const SyncConfig *config, syncstate_t newstate, syncerror_t newSyncError)
+{
+    if (syncConfigs)
+    {
+        assert(config);
+        auto newConfig = *config;
+
+        auto isEnabled = [](syncstate_t state, syncerror_t syncError) -> bool
+        {
+            return state != SYNC_CANCELED && (state != SYNC_DISABLED || syncError != NO_SYNC_ERROR );
+        };
+
+        newConfig.setEnabled(isEnabled(newstate, newSyncError));
+        newConfig.setError(newSyncError);
+
+        syncConfigs->insert(newConfig);
+        return API_OK;
+    }
+    return API_ENOENT;
+}
+
+error MegaClient::changeSyncState(const SyncConfig *config, syncstate_t newstate, syncerror_t newSyncError, bool fireDisableEvent)
+{
+    error e = API_OK;
+    assert(config);
+    if (config)
+    {
+        auto isEnabled = [](syncstate_t state, syncerror_t syncError) -> bool
+        {
+            return state != SYNC_CANCELED && (state != SYNC_DISABLED || syncError != NO_SYNC_ERROR );
+        };
+
+        if ( (config->getError() != newSyncError) || (config->getEnabled() != isEnabled(newstate, newSyncError)) ) //has changed
+        {
+            e = saveAndUpdateSyncConfig(config, newstate, newSyncError);
+        }
+    }
+    else
+    {
+        e = API_ENOENT;
+    }
+
+    app->syncupdate_state(config->getTag(), newstate, newSyncError, fireDisableEvent);
+    return e;
+}
+
+error MegaClient::changeSyncState(int tag, syncstate_t newstate, syncerror_t newSyncError, bool fireDisableEvent)
+{
+    error e = API_OK;
+
+    if (!syncConfigs)
+    {
+        LOG_err << "no SyncConfig upon changeSyncState";
+        e = API_ENOENT;
+    }
+    auto config = syncConfigs?syncConfigs->get(tag):nullptr;
+    assert(config);
+    e = changeSyncState(config, newstate, newSyncError, fireDisableEvent);
+    return e;
+}
+
+error MegaClient::changeSyncStateByNodeHandle(mega::handle nodeHandle, syncstate_t newstate, syncerror_t newSyncError, bool fireDisableEvent)
+{
+    error e = API_OK;
+
+    auto tag = 0;
+
+    if (!syncConfigs)
+    {
+        LOG_err << "no SyncConfig upon changeSyncStateByNodeHandle";
+        e = API_ENOENT;
```

Comment:
Probably we want to return here, rather than applying any further change or trigger callbacks. The action has failed